### PR TITLE
Holographic People

### PR DESCRIPTION
**Pull request by Shadowmech88**
Adds advanced holograms, essentially holographic people.
Advanced holograms are simple mobs with two hands, and hat, uniform, and exosuit inventory slots.
They cannot exist outside of a holodeck.
Ghosts can interact with the holodeck control console in order to be transmogrified into an advanced hologram and placed inside the holodeck. This allows them to play around with the crew in a controlled environment, such as participating in laser tag, combat matches, chess games, and theater performances.

If an advanced hologram attempts to leave the holodeck, it turns back into a ghost. They can also manually turn back into ghosts at any time. Becoming an advanced hologram does not affect any of the ghosts references, meaning they do not lose the ability to re-enter their original body once they turn back into a ghost.

Obviously there are concerns of metagaming. However, the player inside the hologram is given a clear notice upon spawning that they know nothing of the station or its crew except what they see from that point on, and if they were going to choose to ignore that and meta anyway, there already exist easier ways of doing it, such as becoming a mouse and speaking to the AI directly, or even using LOOC to speak to anyone on screen.

Also, throwing code, including throw toggling, has been moved from `/carbon` to `/living`. I didn't notice any misbehavior and I made sure to check for hands so that handless simple_animals and such can't enter throw mode.

:cl:
 * rscadd: Ghosts can now interact with the holodeck control console to be transformed into a holographic person and placed in the holodeck. The player may choose to turn back into a ghost at any time, and can still enter their original body after doing so.

### CODE REVIEW

**Comment by spoonylove** [comment]
emagged holodeck controls should disengage safety protocols.

**Comment by ghost** [comment]
pics?

**Comment by Shadowmech88** [comment]
It just uses the default holographic man sprite that's in the AI's selectable hologram list.

**Comment by small-boss** [comment]
item that allows you to project yourself as a hologram on holopads when

**Comment by Ratvar** [comment]
ai that can project itself as a hologram on holopads WHEN

**Comment by SonixApache** [comment]
this is definitely going to pave the way for holopad memes

**Comment by ghost** [comment]
we can finally have emergency medical holograms on the holodeck
we voyager now

File: code/game/machinery/computer/HolodeckControl.dm [hunk 1]
```diff
@@ -22,6 +22,24 @@
...
 
+/obj/machinery/computer/HolodeckControl/attack_ghost(mob/dead/observer/user)
+	if(!linkedholodeck)
+		return
+	var/turf/spawnturf
+	var/list/L = get_area_turfs(linkedholodeck.type)
+	L = shuffle(L)
+	looping_turfs:
+		for(var/turf/T in L)
+			if(!T.density)
```
Comment (ghost) [inline code/game/machinery/computer/HolodeckControl.dm]: There's a process for this called is_blocked_turf or something like that, so you don't need named loops.

**Comment by ghost** [comment]
I foresee the need for hologrambans.

**Comment by Aurx** [comment]
If they're being a shit as a hologram I don't see how there'd be a need for a roleban instead of just banning them the old fashioned way.

**Comment by ghost** [comment]
Same reasoning as MOMMIbans and cultbans and jobbans in general I guess

**Comment by Aurx** [comment]
MoMMI and cult bans are necessary since other people can try to turn you into them.  
Jobbans are necessary since, short of "we'll ban you if you get picked for this", they're the only way to enforce "Don't be [role]".  
Holobans would stop what, one of half a dozen ways to get back into the round as a ghost?  If somebody's being holoshit they'll be golemshit and dionashit and mouseshit, and posibrainshit and maybe even [l]oocshit too.  A server ban already solves all of those.

**Comment by ghost** [comment]
Fair enough, I forgot about involuntary MoMMIing.

**Comment by Lokiamis** [comment]
At long last - Hologonia rises.

File: code/_onclick/hud/hologram.dm [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	using.icon_state = "act_drop"
+	using.screen_loc = ui_drop_throw
+	using.layer = HUD_BASE_LAYER
+	src.adding += using
+
+	init_hand_icons(ui_style)
+
+	#define ui_swaphand_hologram1	"CENTER-1:16,SOUTH:5"
+	#define ui_swaphand_hologram2	"CENTER:16,SOUTH:5"
+	#define ui_equip_hologram		"CENTER:16,SOUTH:5"
```
Comment (PJB3005) [inline code/_onclick/hud/hologram.dm:88]: You're not cleaning this define up it seems.

File: code/modules/mob/living/simple_animal/hologram.dm [hunk 3]
```diff
@@ -56,4 +56,101 @@
...
+	var/list/obj/abstract/Overlays/obj_overlays[TOTAL_LAYERS]
+
+/mob/living/simple_animal/hologram/advanced/New()
+	..()
+	name = "[name] ([rand(1, 1000)])"
+	real_name = name
+	obj_overlays[HEAD_LAYER]		= getFromPool(/obj/abstract/Overlays/head_layer)
+	obj_overlays[UNIFORM_LAYER]		= getFromPool(/obj/abstract/Overlays/uniform_layer)
+	obj_overlays[SUIT_LAYER]		= getFromPool(/obj/abstract/Overlays/suit_layer)
+	spawn(5)
```
Comment (PJB3005) [inline code/modules/mob/living/simple_animal/hologram.dm]: If the only reason this is behind `spawn()` is to make sure the client receives it, move it to `Login()`.

**Review by PJB3005 — CHANGES_REQUESTED** [review]
Holograms are spread out into multiple files. Give them their own directory.

Also the amount of copy paste from other parts of mob code is horrible, but honestly fixing that would be too big of a refactor to bother with.

**Comment by PJB3005** [comment]
By the way: do people who are holograms still receive "your body is being cloned!" messages? They definitely should.

**Comment by Shadowmech88** [comment]
I feel like I added that feature for transmogrified ghosts in general, but I'm not certain. I'll need to test that.

**Comment by Ephx1** [comment]
We just need Light Bees as an advanced research project now.

File: code/_onclick/hud/hologram.dm (hunk 2 above)
Comment (Shadowmech88) [reply code/_onclick/hud/hologram.dm:88]: Fixed.

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**